### ISSUE TITLE
Singleton/Util/[Factory?] checks should die

### ISSUE THREAD

**Issue by kalessil** [kalessil/phpinspectionsea#317]

**Comment by funivan** [comment]
Can you explain why?

**Comment by Ocramius** [comment]
* `Util` means nothing
 * `Singleton` is basically "how to shoot yourself in the foot 101 (only few rare examples of valid singletons exist, such as an event loop, for example)
 * `Factory` - disagree on this one - unless you want to make everything framework-magic-based, factories are the sanest way to wire together an application (boot-time)

**Comment by afoeder** [comment]
@funivan , to elaborate on Util; better approach is to just simply have a class/an object of everything what your Utility class would provide.

`makeUrlSsl`? --> `(string) (new Uri('http://example.org'))->withScheme('https')`.
`slugString()`? --> `(string) (new Slug('Some arbiträry example'))`

Try it, after a bit of training it's dead simple and you'll love it. Testable as hell!

**Comment by Ocramius** [comment]
Indeed - most `util` class methods can be reduced to single invokable classes. The `util`-class approach typically leads to a single very cluttered god object

**Comment by funivan** [comment]
Thanks. I just misunderstood the title of issue.
Agree at all. Util, Singleton and Factory should be removed from our code.
My IMHO about factory:
`User::createFromRawData($data)`
`User::createFromRawFile($data)`
`User::createFromRepository($id, $repository)`

Maybe this will be better approach then:
```php
$userFactory->withData($data)->build();
```

**Comment by Ocramius** [comment]
Still unclear about the factory one: anyone got examples? My typical factory is like:

```php
final class ConnectionBuilder
{
    public static function connect(array $pdoGarbageParameters) : Connection { ... }
}
```

**Comment by funivan** [comment]
@Ocramius if you can change `Connection` class, then : 
```php
Connection::createFromGarbageParameters($parameters);
Connection::create($ip, $user, $pass)  ; 
```

**Comment by afoeder** [comment]
@Ocramius : y u no
`new ApplicationDbConnection($pdoGarbageParameters)`
while ApplicationDbConnection implements Connection(Interface) and maybe conducts the actual connection on instantiation already.

**Comment by Ocramius** [comment]
> @Ocramius if you can change `Connection` class, then :

Absolutely not - that would introduce a lot of glue code in my service, including type hints against framework-specific details (if I'm getting any information from other infrastructure layers)

 > `new ApplicationDbConnection($pdoGarbageParameters)`

Same as above - now my procedural code landed in a class, and I mixed procedural and OO code.

To be clear: factories *ARE* procedural code, and there's no way around it. The typical scenario is:

 * 1. read some config from somewhere 
 * 2. check the current system (maybe you are dumb enough to deploy something on windows)
 * 3. actually build the thing that is needed

This is stuff that you don't want to land in your runtime objects, but is to happen *BEFORE* the application runs, and only rarely (lazily) during app run.

**Comment by funivan** [comment]
@Ocramius How about ConnectionParametersInterface. ?
Then you can have procedural code in `GarbageParamerers` instead of Factory.

**Comment by Ocramius** [comment]
@funivan that just hides the procedural code even more by adding layers of indirection that make it even more confusing to use: you still have to take decisions on what/how to instantiate something at some point.

**Comment by funivan** [comment]
@Ocramius I get your point. Word "factoy" can be indicator of some hacks (in some cases)
But, maybe it would be nice to include this in the inspection. For example if we can implement logic in the class `Connection::__construct` and this logic is pretty clear, nice and not procedural.  
We will double check our factory. If we really want this class, we can suppress inspection. But in other ways, we will always create nice constructors. What do you think?

**Comment by Ocramius** [comment]
> For example if we can implement logic in the class Connection::__construct and this logic is pretty clear, nice and not procedural.

That makes it absolutely terrible to test your `Connection::__construct()` afterwards, and instantiating your object may cause side effects. 

Rejecting values in `__construct` can be OK, but messing around with them is a bad smell.

 > We will double check our factory. 

I implement loads of factories to support my DIC, for example - [ZF even has a simple interface for it](https://github.com/zendframework/zend-servicemanager/blob/fc81deea9f6767dfbed549aa3f168d5977651d20/src/Factory/FactoryInterface.php)

 > But in other ways, we will always create nice constructors.

A nice constructor just accepts what is given to it, and eventually throws an exception if it doesn't like it, but it shouldn't mess with what is given as input: that's where you cut the line between constructor and factory.

**Comment by davidmpaz** [comment]
+1 for:
> Factory - disagree on this one - unless you want to make everything framework-magic-based, factories are the sanest way to wire together an application (boot-time)

At some point you need/want to get procedural. Nothing wrong with that. Factories can help to find the balance between procedural and OO at that time.

**Comment by kalessil** [comment]
For future people: 

- Class violates singleton/factory pattern definition: dropped
- Private constructor semantics: dropped

Not all going to agree, but this for best.
